# TIL6022 - Final Assignment
Members: Joris Voogt, Xiaoyu Mao, Yuwen Peng  
Student numbers: 4295978, 5947642, 5914507

# Research Objective

*Requires data modeling and quantitative research in Transport, Infrastructure & Logistics*

- RQ: GDP is used to represent the economic development of countries, on the basis of which the changes in the scale of road and rail passenger transport in European countries in recent years and its impact on road traffic safety are analysed.

This research will focus on passenger transport trends and safety issues in Europe. We will analyse the relationship between transport mode choice and the economy in different countries.
The GDP is used to represent the economic development of the country, on the basis of which we will analyse the changes in the scale of road and rail passenger transport in European countries in recent years. By analyzing the volume of rail passenger and road passenger transport ,we can find the proportion of both in the overall volume of transport. 
Then we can discuss whether there are countries who are highly dependent on road transport, and then we select the countries who have a large proportion of road transport to analyze the relationship between GDP and traffic safety.

Time scale: Data is a bit limited for passenger transport, so most likely in the range of 2016-2021.
Geographical boundary: Countries in Europe. Not all of them have data available in the datasets, so countries with most data available will be analysed.
 
A short list of how we plan to do this analysis:

1. The volume of passenger transport relative to GDP can give the relative scale of passenger transport in the national economy. Visualise it by creating an interactive map of Europe with different colors for each country based on the data, representing levels of the volume of passenger transport relative to GDP for different transport modes.

2.  Analyse the changes in the scale of road and rail passenger transport in European countries in recent years. By analysing the volume of rail passenger transport and road passenger transport in each country and drawing a scale diagram, we can find out the proportion of the two in the overall volume of transport; and drawing a line graph can show the changes in recent years.

3.  Further explore the impact of transport investment on transport safety. Discuss whether there are countries in 2 that are highly dependent on road transport, and then select countries with a large share of road transport to analyse the relationship between GDP and traffic safety.


4. Look for other possible factors that could be related to traffic mortality rate. Such as proportions of motorways, GDP per capita, alcohol consumption. This data is likely leading to new relationships and possibly correlates. Therefore, a correlation matrix is used, measuring strength and quickly giving an overview of all the numerical correlations.

5. To more scientifically evaluate the relationship between traffic mortality rate and other factors, it would be better to establish a regression model. The result obtained will tell us whether there is a positive or negative relationship between them. (if we have enough time) 

6. Summarize and propose ideas about transport patterns in the future and strategize about road safety.

In [1]:
# import libraries
import pandas as pd
import numpy as np
import math
import matplotlib.pyplot as plt
import plotly.express as px
import seaborn as sns

# Part I - Data Import
First, we are going to create new data frame, clean the useless columns and rows from the original data, then rename the data columns and rows to be more clearly for the following data:
 - Volume of passenger transport relative to GDP
 - Rail transport of passengers
 - Road transport of passengers
 - Road Infrustructure Maintenance Spending
 - Road injury crashes, fatalities and injuries
 - Road fatalities per one hundrend thousand inhabitants
 - GDP per capita

I'm starting off with the volume of passenger transport relative to GDP.

The indicator about this data is defined as the ratio between the wolume of inland passenger transport measured in passenger-kilometers and GDP.It includes transport on national territory by passenger car, bus, coach and train.

In [2]:
# Import Volume of passenger transport relative to GDP: Read the csv file as a dataframe
file_path_pg_GDP = 'data/pg_GDP.csv'
df_pg_GDP = pd.read_csv(file_path_pg_GDP)

# Rename columns and rows
df_pg_GDP.columns = df_pg_GDP.iloc[6].tolist()
df_pg_GDP = df_pg_GDP.rename(columns={df_pg_GDP.columns[0]: 'GEO'})

# Clean the data by dropping columns of the dataframe that will not be used
# These rows are used to indicate the data, and columns are used to explain the accuracy of the data
df_pg_GDP = df_pg_GDP.iloc[8:48]
columns_to_remove = [2,4,6,8,10,12,14,16,18,20,22,24]
df_pg_GDP = df_pg_GDP.drop(df_pg_GDP.columns[columns_to_remove], axis=1 )
columns_to_remove = [7]
df_pg_GDP = df_pg_GDP.drop(df_pg_GDP.columns[columns_to_remove], axis=1 )

# Delete the rows without data and use Euro area or country as the row name
df_pg_GDP = df_pg_GDP.set_index('GEO')
df_pg_GDP = df_pg_GDP[~(df_pg_GDP == ':').all(axis=1)]

df_pg_GDP.head() # disaplay the first 5 rows of dataframe

2010   2011   2012   2013   2014  \
GEO                                                                            
European Union - 27 countries (from 2020)  103.7  102.2  100.9  100.8  100.1   
Belgium                                    114.1  110.5    114  111.5  103.7   
Bulgaria                                    91.5   91.2   91.9   94.4   99.1   
Czechia                                    101.2    100   99.6  100.7  103.1   
Denmark                                     98.7   99.5   98.7   97.7   98.1   

                                          2015   2016  2017   2018   2019  \
GEO                                                                         
European Union - 27 countries (from 2020)  100    100  98.3   96.6   95.4   
Belgium                                    100   97.4  96.6   95.5   93.8   
Bulgaria                                   100   95.9  93.9   89.7   88.7   
Czechia                                    100  100.7  99.2  100.9  101.5   
Denmark                                    100   99.8  98.6   97.2   98.2   

                                           2020  2021  
GEO                                                    
European Union - 27 countries (from 2020)  79.2  80.2  
Belgium                                    80.9  79.9  
Bulgaria                                   75.4  72.3  
Czechia                                    83.8  99.2  
Denmark                                    88.4  85.9

In [22]:
# Import Rail transport of passengers: Read the csv file as a dataframe
file_path_rail_pg = 'data/rail_pg.csv'
df_rail_pg = pd.read_csv(file_path_rail_pg)

# Rename columns and rows
df_rail_pg.columns = df_rail_pg.iloc[7].tolist()
df_rail_pg = df_rail_pg.rename(columns={df_rail_pg.columns[0]: 'GEO'})

# Clean the data by dropping columns of the dataframe that will not be used
df_rail_pg = df_rail_pg.iloc[9:49]
columns_to_remove = [2,4,6,8,10,12,14,16,18,20,22,24]
df_rail_pg = df_rail_pg.drop( df_rail_pg.columns[columns_to_remove], axis=1 )

# Delete the rows without data and reset the index
df_rail_pg = df_rail_pg[~(df_rail_pg == ':').all(axis=1)]
df_rail_pg = df_rail_pg.reset_index(drop = True)

df_rail_pg.head()

GEO    2011   2012   2013   2014  \
0  European Union - 27 countries (from 2020)       :      :      :      :   
1       Euro area – 20 countries (from 2023)       :      :      :      :   
2      Euro area - 19 countries  (2015-2022)       :      :      :      :   
3                                    Belgium  10,498      :      :      :   
4                                   Bulgaria   2,059  1,870  1,821  1,698   

      2015     2016     2017     2018     2019     2020     2021   2022  
0  375,713  384,335  394,142  400,427  413,923  223,668  260,715      :  
1        :        :        :        :        :        :        :      :  
2        :        :        :        :        :        :        :      :  
3        :        :        :        :        :        :        :      :  
4    1,549    1,455    1,434    1,476    1,520    1,118    1,203  1,600

In [25]:
# Import Road transport of passengers: Read the csv file as a dataframe
file_path_road_pg = 'data/road_pg.csv'
df_road_pg = pd.read_csv(file_path_road_pg)

# Rename columns and rows
df_road_pg.columns = df_road_pg.iloc[7].tolist()
df_road_pg = df_road_pg.rename(columns={df_road_pg.columns[0]: 'GEO'})

# Clean the data by dropping columns of the dataframe that will not be used
# After observation, it was found that they are all sequence numbers with double digits
# Remove double numbered columns starting from the fourth column with index 3
df_road_pg = df_road_pg.iloc[9:37]
columns_to_keep = df_road_pg.columns[:1].to_list()
columns_to_remove = df_road_pg.columns[1::2]
df_road_pg = df_road_pg[columns_to_keep + columns_to_remove.to_list()]

# Delete the rows without data and reset the index
df_road_pg = df_road_pg[~(df_road_pg == ':').all(axis=1)]
df_road_pg = df_road_pg.reset_index(drop = True)

df_road_pg.head()

GEO        2013        2014        2015        2016    2017  \
0   Belgium      15,452      15,215      14,201      13,306       :   
1  Bulgaria  10,316.800  11,447.100  12,257.100  12,210.200  10,553   
2   Czechia      14,703  15,750.800  15,406.910  15,841.300  16,846   
3   Denmark           :           :           :           :       :   
4   Germany      60,479      62,227      65,075      64,073  62,481   

        2018        2019       2020       2021  
0          :           :          :          :  
1  9,886.500  10,835.700  5,037.600  4,949.700  
2     17,265      17,056  8,593.400  8,351.400  
3          :           :          :          :  
4     62,520      61,186     34,025     34,259

In [5]:
# Import Road Infrastructure Investment Spending: Read the csv file as a dataframe
file_path_road_inv = 'data/road_inv.csv'
df_road_inv = pd.read_csv(file_path_road_inv)

# First clean the data by dropping columns of the dataframe that will not be used
df_road_inv = df_road_inv.iloc[4:56]
# Remove double numbered columns starting from the fourth column with index 3
columns_to_keep = df_road_inv.columns[:1].to_list()
columns_to_remove = df_road_inv.columns[3::2]
df_road_inv = df_road_inv[columns_to_keep + columns_to_remove.to_list()]

# Rename columns and rows
new_column_names = [str(year) for year in range(1999, 2022)]
df_road_inv.columns = new_column_names
df_road_inv = df_road_inv.rename(columns={df_road_inv.columns[0]: 'GEO'})

# Delete the rows without data and reset the index
df_road_inv = df_road_inv[~(df_road_inv == '..').all(axis=1)]
df_road_inv = df_road_inv.reset_index(drop = True)

df_road_inv.head()

GEO           2000           2001           2002           2003  \
0     Albania    108,347,272    107,567,894     64,458,009     60,782,802   
1     Armenia             ..      1,371,881        830,439        418,066   
2   Australia  3,985,384,644  3,232,914,774  3,174,141,910  3,809,467,524   
3     Austria    475,000,000    640,000,000    532,000,000    650,000,000   
4  Azerbaijan     30,255,355     47,961,631     46,789,989     34,212,659   

            2004           2005           2006           2007           2008  \
0     91,078,262     68,193,896    175,223,508    253,261,449    499,989,410   
1      3,923,281     19,854,465     23,142,419     29,057,065     50,248,397   
2  4,673,112,063  6,047,684,590  6,696,458,225  7,813,640,359  7,904,953,310   
3    720,000,000    687,000,000    802,000,000    870,000,000    875,000,000   
4     48,289,409     82,349,945    260,388,800    374,213,302  1,328,480,437   

   ...            2012            2013            2014            2015  \
0  ...     180,820,940     234,163,844     192,718,553     179,236,201   
1  ...      26,502,499      23,186,815      66,773,299      77,721,547   
2  ...  15,649,722,083  13,990,189,647  10,988,504,646  10,310,609,857   
3  ...     327,000,000     363,000,000     453,000,000     455,000,000   
4  ...   1,479,179,001   1,913,627,639   1,411,301,928     873,204,688   

             2016            2017            2018            2019  \
0      89,140,344     159,032,880     169,252,490     175,879,111   
1      90,425,117              ..              ..              ..   
2  10,841,616,558  12,673,185,564  14,363,912,675  13,317,666,734   
3     444,000,000     515,000,000     463,000,000     562,000,000   
4     498,158,118     557,071,513     695,011,451     844,323,018   

             2020            2021  
0     223,292,017     265,726,260  
1              ..              ..  
2  12,671,465,384  13,286,823,655  
3     548,000,000     480,000,000  
4     757,138,439   1,208,250,497  

[5 rows x 23 columns]

In [6]:
# Import Road Infrastructure Maintenance Spending: Read the csv file as a dataframe
file_path_road_maint = 'data/road_maint.csv'
df_road_maint = pd.read_csv(file_path_road_maint)

# Rename columns and rows
df_road_maint.columns = df_road_maint.iloc[2].tolist()
df_road_maint = df_road_maint.rename(columns={df_road_maint.columns[0]: 'GEO'})

# Clean the data by dropping columns of the unit that is meaningless when the selected country are all from Europe
# Then remove the rows from original data for explaination
df_road_maint = df_road_maint.iloc[4:46]
columns_remove = [1]
df_road_maint = df_road_maint.drop( df_road_maint.columns[columns_remove], axis=1 )

# Delete the rows without data and reset the index
df_road_maint = df_road_maint[~(df_road_maint == '..').all(axis=1)]
df_road_maint = df_road_maint.reset_index(drop = True)

df_road_maint.head()

GEO         2000         2001         2002         2003  \
0     Albania    3,838,572    5,705,714    6,824,432    5,767,021   
1     Armenia           ..           ..           ..           ..   
2     Austria  508,000,000  520,000,000  294,000,000  388,000,000   
3  Azerbaijan           ..   15,587,530   15,233,950   17,106,329   
4     Belgium   59,000,000   62,000,000   68,000,000   73,000,000   

          2004         2005         2006         2007         2008  ...  \
0    6,829,499    7,110,073    5,670,979    6,046,008    8,056,200  ...   
1           ..    8,770,105    9,588,976   10,683,530   11,111,985  ...   
2  458,000,000  443,000,000  495,000,000  486,000,000  467,000,000  ...   
3   34,375,512   33,109,772   54,396,290   31,467,937   34,742,328  ...   
4   77,000,000   80,000,000  104,000,000   94,000,000  102,000,000  ...   

          2012         2013         2014         2015         2016  \
0    6,652,853    8,745,595   15,262,572    8,374,520   13,004,791   
1   10,706,482   10,085,012   10,050,491   11,236,616   11,586,347   
2  517,000,000  559,000,000  667,000,000  692,000,000  697,000,000   
3   34,537,201   31,669,866   31,660,750   22,909,507   18,702,182   
4  145,000,000  147,000,000  206,000,000  457,000,000  528,000,000   

          2017         2018         2019         2020         2021  
0   13,593,700   13,444,559   13,050,669    9,058,683   21,515,139  
1           ..           ..           ..           ..           ..  
2  687,000,000  726,000,000  752,000,000  791,000,000  899,000,000  
3   27,802,090   25,390,819   32,049,598   52,056,489   46,719,682  
4  396,797,000  216,000,000  213,000,000  155,000,000  498,000,000  

[5 rows x 23 columns]

In [9]:
# Import Road injury crashes, fatalities and injuries: Read the csv file as a dataframe
df_inj_kill = pd.read_csv('data/road_fat.csv')

# First clean the data by dropping columns of the dataframe that will not be used
df_inj_kill = df_inj_kill.iloc[4:60]
# Remove double numbered columns starting from the fourth column with index 3
columns_to_keep = df_inj_kill.columns[:1].to_list()
columns_to_remove = df_inj_kill.columns[3::2]
df_inj_kill = df_inj_kill[columns_to_keep + columns_to_remove.to_list()]

# Rename columns and rows
new_column_names = [str(year) for year in range(1999, 2022)]
df_inj_kill.columns = new_column_names
df_inj_kill = df_inj_kill.rename(columns={df_inj_kill.columns[0]: 'GEO'})

df_inj_kill = df_inj_kill.reset_index(drop = True)

df_inj_kill.head()

GEO    2000    2001    2002    2003    2004    2005    2006    2007  \
0    Albania     620     547     478     510   1,110   1,180   1,342   1,728   
1  Argentina      ..      ..      ..      ..      ..      ..      ..      ..   
2    Armenia   1,377   1,495   1,448   1,546   1,751   2,084   2,421   3,091   
3  Australia  28,591  29,066  29,483  29,891  30,330  32,069  33,740  34,005   
4    Austria  55,905  57,223  57,640  57,812  56,735  54,002  52,660  53,902   

     2008  ...     2012     2013    2014    2015    2016     2017    2018  \
0   1,554  ...    2,569    2,798   2,617   2,692   2,779    2,611   2,291   
1      ..  ...  122,062  118,925  85,984      ..      ..  118,593      ..   
2   3,532  ...    4,050    4,310   4,776   5,084   4,718    5,458      ..   
3  34,839  ...   35,367   36,322  36,703  38,286  40,237   40,553  40,732   
4  51,200  ...   51,426   48,499  48,100  47,845  48,825   47,672  46,934   

     2019    2020    2021  
0   2,044   1,598   1,860  
1      ..      ..      ..  
2      ..      ..      ..  
3      ..      ..      ..  
4  45,556  38,074  41,251  

[5 rows x 23 columns]

In [10]:
# Import Road fatalities per one hundred thousand inhabitants: Read the csv file as a dataframe
file_path_fat_inh = 'data/fat_inh.csv'
df_fat_inh = pd.read_csv(file_path_fat_inh)

# First clean the data by dropping columns of the dataframe that will not be used
# These rows are used to indicate the data, and columns are used to explain the accuracy of the data
df_fat_inh = df_fat_inh.iloc[3:62]

# Remove double numbered columns starting from the fourth column with index 3
columns_to_keep = df_fat_inh.columns[:1].to_list()
columns_to_remove = df_fat_inh.columns[2::2]
df_fat_inh = df_fat_inh[columns_to_keep + columns_to_remove.to_list()]

# Rename columns and rows
new_column_names = [str(year) for year in range(1999, 2022)]
df_fat_inh.columns = new_column_names
df_fat_inh = df_fat_inh.rename(columns={df_fat_inh.columns[0]: 'GEO'})
# Reset the index
df_fat_inh = df_fat_inh.reset_index(drop = True)

df_fat_inh.head()

GEO  2000  2001  2002  2003  2004  2005 2006  2007  2008  ...  2012  \
0    Albania   9.1   9.7   8.2   8.6  10.4  10.2  9.3  12.9  10.3  ...  11.5   
1  Argentina    ..    ..    ..    ..    ..    ..   ..    ..  14.3  ...  12.2   
2    Armenia   6.8   7.6   7.6   8.2   8.4  10.2   11  12.3  13.6  ...  10.7   
3  Australia   8.6   8.2   7.8   7.3   7.2   7.3  7.1     7   6.2  ...   5.6   
4    Austria  12.2  11.9  11.8  11.5  10.7   9.3  8.8   8.3   8.2  ...   6.3   

   2013  2014 2015  2016  2017 2018  2019 2020 2021  
0  10.2   9.1  9.4   9.4   7.7  7.4     8  6.4    7  
1  12.3  12.4   ..  12.8  12.3   ..    ..   ..   ..  
2  10.9  10.3   12   9.3   9.8   ..    ..   ..   ..  
3   5.5   4.9  5.1   5.3     5  4.5   4.7  4.3  4.3  
4   5.4     5  5.5   4.9   4.7  4.6   4.7  3.9    4  

[5 rows x 23 columns]

- Add more data to analyse the factors that possibly related to safety
Alcohol consumption from WHO only collected from 2010 to 2019, so the other data also need to be capture to this period.

In [104]:
# Import Proportion of motorways over the total road network(%),2010-2021
df_mot_rate = pd.read_csv('data/mot_rate.csv')
df_mot_rate = df_mot_rate.iloc[:,:11]
df_mot_rate

# Import Motorization index(in cars per 1000 inhabitants),2000-2021
df_mot_index = pd.read_csv('data/mot_index.csv')
df_mot_index = df_mot_index.iloc[:,0:1].join(df_mot_index.iloc[:,11:21],how='outer')
df_mot_index

# Import Unemployment rate(%),2010-2022
df_unempl_rate = pd.read_csv('data/unempl_rate.csv')
df_unempl_rate = df_unempl_rate.iloc[:,:11]
df_unempl_rate

# Import alcohol consumption(%),2010-2021
df_alcohol = pd.read_csv('data/alcohol.csv')
df_alcohol = df_alcohol.iloc[:,0:1].join(df_alcohol.iloc[:,11:],how='outer')
df_alcohol

GEO   2010   2011   2012  \
0                                             Albania   4.88   5.03   4.43   
1                                             Andorra  10.53  10.46  10.33   
2                                             Armenia   4.23   4.07   3.89   
3                                             Austria  12.10  11.90  12.10   
4                                          Azerbaijan   0.99   0.60   0.52   
5                                             Belarus  14.43  14.50  13.56   
6                                             Belgium  10.27  10.14  10.09   
7                              Bosnia and Herzegovina   4.57   4.71   4.83   
8                                            Bulgaria   9.82   9.88  10.88   
9                                             Croatia  10.71  11.01  10.58   
10                                             Cyprus  11.33  10.71  10.58   
11                                            Czechia  12.68  12.56  12.83   
12                                            Denmark  10.27  10.45   9.23   
13                                            Estonia  14.97  16.27  16.96   
14                                            Finland   9.72   9.81   9.27   
15                                             France  12.30  12.35  12.22   
16                                            Georgia   7.72   8.29   7.23   
17                                            Germany  11.58  11.33  11.32   
18                                             Greece   8.33   7.48   7.62   
19                                            Hungary  10.70  11.37  11.13   
20                                            Iceland   6.79   6.84   6.82   
21                                            Ireland  11.57  11.49  11.38   
22                                             Israel   2.78   2.81   2.89   
23                                              Italy   6.95   6.98   7.48   
24                                         Kazakhstan   3.94   3.41   3.73   
25                                         Kyrgyzstan   6.02   5.95   5.98   
26                                             Latvia   9.83  10.11  10.20   
27                                          Lithuania  13.22  14.76  14.97   
28                                         Luxembourg  11.84  11.74  11.67   
29                                         Montenegro   9.94   9.02   9.37   
30                                        Netherlands   9.34   8.95   9.00   
31                                             Norway   6.59   6.44   6.21   
32                                             Poland  10.07  10.24  10.21   
33                                           Portugal  11.26  10.86   9.76   
34                                            Romania  10.01   9.58   9.66   
35                                 Russian Federation  10.82  10.75  10.56   
36                                             Serbia   8.87   8.55   8.78   
37                                           Slovakia  10.16  10.27  10.14   
38                                           Slovenia  10.33  10.61  10.95   
39                                              Spain   8.94   8.74   8.60   
40                                             Sweden   7.38   7.31   7.23   
41                                        Switzerland  10.01   9.99   9.87   
42                                            Türkiye   1.42   1.47   1.47   
43                                         Tajikistan   0.68   0.71   0.77   
44                                       Turkmenistan   2.90   2.88   3.00   
45                                            Ukraine   7.77   8.74   8.54   
46  United Kingdom of Great Britain and Northern I...  10.01   9.80   9.53   
47                                         Uzbekistan   1.74   1.73   1.95   

     2013   2014   2015   2016   2017   2018   2019  
0    4.28   4.40   4.33   4.38   4.39   4.44   4.40  
1   10.06   9.88  11.02  11.18  11.18  10.72  10.99  
2    3.92   4.22   4.04   3.83   3.83   3.69   3.77  
3   12.10  

In [11]:
# Import GDP per capita: Read the csv file as a dataframe
file_path_GDP = 'data/GDP.csv'
df_GDP = pd.read_csv(file_path_GDP)

# Remove the last row and use European region or country as the row name
df_GDP = df_GDP.iloc[:58]
df_GDP = df_GDP.rename(columns={df_GDP.columns[0]: 'GEO'})
#df_GDP = df_GDP.set_index('GEO')

df_GDP.head()

GEO       2000       2001       2002       2003       2004  \
0     Albania   1,126.70   1,281.70   1,425.10   1,846.10   2,373.60   
1     Andorra  21,620.50  22,809.10  24,785.00  31,954.20  37,630.80   
2     Armenia      603.3      676.2      765.3      910.2   1,166.60   
3     Austria  24,625.60  24,558.80  26,527.60  32,294.00  36,889.20   
4  Azerbaijan      655.1      703.7      763.1      883.7   1,045.00   

        2005       2006       2007       2008  ...       2013       2014  \
0   2,673.80   2,972.70   3,595.00   4,370.50  ...   4,413.10   4,578.60   
1  39,583.90  43,084.30  50,562.90  53,721.40  ...  44,747.80  45,680.50   
2   1,608.20   2,109.50   3,064.30   3,908.90  ...   3,833.20   4,017.20   
3  38,417.50  40,669.30  46,915.30  51,920.00  ...  50,731.10  51,786.40   
4   1,578.40   2,473.00   3,851.30   5,574.50  ...   7,875.30   7,890.80   

        2015       2016       2017       2018       2019       2020  \
0   3,952.80   4,124.10   4,531.00   5,287.70   5,396.20   5,343.00   
1  38,885.50  39,931.20  40,632.20  42,904.80  41,328.60  37,207.20   
2   3,666.10   3,680.00   4,042.00   4,391.90   4,828.50   4,505.90   
3  44,195.80  45,307.60  47,429.20  51,466.60  50,070.40  48,809.20   
4   5,500.50   3,880.70   4,147.20   4,739.80   4,805.80   4,229.90   

        2021       2022  
0   6,377.20   6,802.80  
1  42,072.30  41,992.80  
2   4,966.50   7,014.20  
3  53,637.70  52,131.40  
4   5,408.00   7,736.70  

[5 rows x 24 columns]

# Part II - Data processing

- This part is preparation for the subsequent quantitative analysis and data visualization.

Divide countries based upon GDP:

In [12]:
# Divided the countries into two groups: low-income and high-income countries, basing on GDP per capita
# Select the last five years as evaluation criteria
selected_years = ['2018','2019','2020','2021','2022']
data_selected_years = df_GDP[selected_years]

# Replace invalid values in the dataframe with NAN
data_selected_years = data_selected_years.replace('..', np.nan)
# Convert numerical values to floats
data_selected_years = data_selected_years.replace(',', '', regex=True).astype(float)

# Calculate the average GDP of each country
data_average_GDP = data_selected_years.mean(axis=1)
df_GDP['average GDP'] = data_average_GDP

# Calculate the average GDP of all countries
average_GDP = df_GDP['average GDP'].mean()
print(f'The average GDP of all the countries in last five years is {average_GDP}')
# Add a new column to mark their level of income: high or low
df_GDP['Income Group'] = ''
df_GDP.loc[df_GDP['average GDP'] > average_GDP,'Income Group'] = 'High Income'
df_GDP.loc[df_GDP['average GDP'] < average_GDP,'Income Group'] = 'Low Income'

# Create lists to devide the two groups of countries
High_inc = df_GDP[df_GDP['Income Group'] == 'High Income']['GEO'].to_list()
Low_inc = df_GDP[df_GDP['Income Group'] == 'Low Income']['GEO'].to_list()

# Print two lists of countries one by one
if True:
    print('\033[1mHigh Income countries includes:\033[0m')
    for country in High_inc:
        print(country, end = ',')
    print('\n\033[1mLow Income countries includes:\033[0m')
    for country in Low_inc:
        print(country, end = ',')

The average GDP of all the countries in last five years is 36946.38868421052
High Income countries includes:
Andorra,Austria,Belgium,Channel Islands,Denmark,Faroe Islands,Finland,France,Germany,Greenland,Iceland,Ireland,Isle of Man,Liechtenstein,Luxembourg,Monaco,Netherlands,Norway,San Marino,Sweden,Switzerland,United Kingdom,
Low Income countries includes:
Albania,Armenia,Azerbaijan,Belarus,Bosnia and Herzegovina,Bulgaria,Croatia,Cyprus,Czechia,Estonia,Georgia,Greece,Hungary,Italy,Kazakhstan,Kosovo,Kyrgyz Republic,Latvia,Lithuania,Moldova,Montenegro,North Macedonia,Poland,Portugal,Romania,Russian Federation,Serbia,Slovak Republic,Slovenia,Spain,Tajikistan,Turkiye,Turkmenistan,Ukraine,Uzbekistan,

Create dataframe which contains the percentages of passenger kilometres for both road (buses and coaches) and rail.

In [91]:
# The years we choose to analyze (based upon availability)
chosen_years = [str(i) for i in range(2013, 2022)]
# Keep the country column
chosen_years.insert(0, 'GEO')

# The countries we choose to analyze (based upon availability)
chosen_countries = ('Bulgaria', 'Czechia', 'Denmark', 'Germany', 'France', 'Croatia', 'Italy', 'Lithuania', 'Austria', 'Poland', 'Slovakia', 'Finland')

# Keep only the chosen years and countries in the road and rail passenger km data sets
df_road = df_road_pg[df_road_pg['GEO'].isin(chosen_countries)][chosen_years]
df_road.reset_index(drop=True, inplace=True)
df_rail = df_rail_pg[df_rail_pg['GEO'].isin(chosen_countries)][chosen_years]
df_rail.reset_index(drop=True, inplace=True)

# Combine road and rail data
df_psk = pd.DataFrame(columns=['Year', 'Country', 'Rail_psk', 'Road_psk', 'Total_psk', 'Percentage road', 'Percentage rail'])

for row in range(0, len(df_road.index)):  
    for col in range(1, len(df_road.columns)):
        
        # Remove commas and change : to NaN
        rail_val = float(df_rail.iloc[row, col].replace(',', '').replace(':', 'NaN'))
        road_val = float(df_road.iloc[row, col].replace(',', '').replace(':', 'NaN'))
        
        df_psk.loc[len(df_psk)] = [df_road.columns[col], 
                                   df_road.iloc[row, 0],
                                   rail_val,
                                   road_val,
                                   rail_val+road_val,
                                   road_val/(rail_val+road_val)*100,
                                   rail_val/(rail_val+road_val)*100
                                   ]

# Show dataframe
df_psk

Year   Country  Rail_psk  Road_psk  Total_psk  Percentage road  \
0   2013  Bulgaria    1821.0   10316.8    12137.8        84.997281   
1   2014  Bulgaria    1698.0   11447.1    13145.1        87.082639   
2   2015  Bulgaria    1549.0   12257.1    13806.1        88.780322   
3   2016  Bulgaria    1455.0   12210.2    13665.2        89.352516   
4   2017  Bulgaria    1434.0   10553.0    11987.0        88.037040   
..   ...       ...       ...       ...        ...              ...   
94  2017   Finland    3754.0       NaN        NaN              NaN   
95  2018   Finland    3792.0       NaN        NaN              NaN   
96  2019   Finland    3957.0       NaN        NaN              NaN   
97  2020   Finland    2133.0    2589.0     4722.0        54.828463   
98  2021   Finland    1969.0       NaN        NaN              NaN   

    Percentage rail  
0         15.002719  
1         12.917361  
2         11.219678  
3         10.647484  
4         11.962960  
..              ...  
94              NaN  
95              NaN  
96              NaN  
97        45.171537  
98              NaN  

[99 rows x 7 columns]

# Part III - Data Visualization

# Eurostat to Map data
- This next block of code transforms the volume of passengers relative to GDP data to a dataframe which can be used in a choropleth map. The color variations representing different levels of the data. 
- A gradient color spectrum can be used so that countries with high ratios are shown in darker colors and countries with low ratios are shown in lighter colors.
- Add a timeline at the bottom or side of the map to represent the time range. Users can drag the timeline or click on different years to see the Volume of passenger transport relative to GDP for each country at different points in time.
- Parts of this can be reused for other eurostat data sets.

# Choropleth Map using Plotly
This next block of code creates a map of Europe using plotly's choropleth.
In this particular instance, it plots transport volumes relative to GDP.
Most of this code can be reused for plotting other data sets using choropleth.

In [85]:
# Read in prepared data set, rename a column and extract the different years
df = pd.read_csv('data/pg_GDP_map.csv')
df = df.rename(columns={'pg_GDP': 'index'})
years = df.year.unique()

# Calculate max value for continuous colour range
max_pg_gdp = df['index'].max()

# Country names to be used in the labeling
customdata = df['country']

# Create choropleth figure
fig = px.choropleth(df,
                    locations='iso_alpha',  # Iso-alpha-3 codes to signify which country
                    color='index',
                    hover_name='country',
                    color_continuous_scale=[[0, 'gray'], [0.01, 'gray'], [0.01, 'blue'], [1, 'red']],  # Gray represents countries with no available data.
                    projection='miller',
                    range_color=(0,max_pg_gdp),
                    scope='europe',
                    animation_frame='year',
                    animation_group='country'
                    )

# Sets hover data
fig.update_traces(customdata=customdata, 
                  hovertemplate= np.select([df['index'] == 0], ['<b>Country: </b> %{customdata}<br><br>No data available'], '<b>Country: </b>%{customdata}<br><br><b>Index value: </b>%{z}'))

# Gets value for the entire EU
numb = df.query('country=="EU"')['index']

# Get current year
current_year = fig.layout.sliders[0].steps[0].label

# Set title and EU index and legend position
fig.update_layout(title='<b>Volume of passengers relative to GDP</b><br>EU index: ' + str(numb.tolist()[0]),
                   coloraxis_colorbar=dict(x=0,
                                           y=0.5)
                  )

# Set figure size
fig.update_layout(width=1000,
                  height=600)

fig.show()

"Volume of passenger transport relative to GDP" is a key economic indicator that provides insights into the relationship between the volume of passenger transportation servicesand the Gross Domestic Product (GDP) of a country or region.  
A high volume of passenger transport relative to GDP often indicates a robust and growing economy. When the volume of passenger transport is significant in comparison to the GDP, it suggests that the transportation sector is making a substantial contribution to economic growth.   
A high ratio may suggest that a country's economy relies heavily on transportation services, such as tourism, logistics, or commuting. In this case, fluctuations in transportation services may have a substantial impact on the overall economic health of the country.    
A low ratio relative to GDP might suggest that a country or region's transportation system is highly efficient. In such cases, the transportation sector doesn't need to consume a significant portion of the economic output to move people efficiently, indicating well-developed and cost-effective transport infrastructure.  
- The data on the "Volume of passenger transport relative to GDP" in the European Union from 2010 to 2022, with 2015 defined as 100, can be analyzed as follows:
1. Positive Trend:
    From 2010 to 2015, the ratio increased from 103.7 to 100, where 2015 was defined as 100. This indicates a relative increase in the volume of passenger transport in relation to GDP during this period. The passenger transport sector was playing an increasingly significant role in the European economy. This have been due to factors such as increased tourism, or infrastructure development and investment.
2. Stability:
    From 2016 to 2018, the ratio remained relatively stable, hovering around 100. This suggests that the passenger transport sector was maintaining its level of contribution to the economy, with no significant shifts in either direction.
3. Decline:
    From 2019 onwards, there was a notable decline. By 2022, the ratio had decreased to 80.2. This decline indicates a reduction in the relative contribution of the passenger transport sector to the GDP. This reduction could be influenced by various factors, such as changes in travel behavior (e.g.due to the COVID-19 pandemic), economic shifts, or increased efficiency in the transportation sector. No data augmentation after the pandemicmay because many people adapted to remote work and virtual meetings. As more people work from home, the need for daily commuting and business travel may decrease, impacting the passenger transport sector. Many businesses in the travel and tourism sectors suffered financial losses.Reduced business travel and tourism can lead to lower passenger transport relative to GDP.

# Percentage road and rail passenger kilometres
Next, we look at the percentage of total passenger kilometres that is road (buses and coaches) and rail kilometres.
We do this by plotting the dataframe created in part II as a bar chart with animation frames for the years.

In [98]:
# Remove Denmark as it has no entries for road_psk
df_psk_bar = df_psk.drop(df_psk[df_psk['Country'] == 'Denmark'].index)

# Create bar chart with animation frames for years
fig_bar = px.bar(df_psk_bar,
                 x='Country',
                 y=['Percentage road', 'Percentage rail'],
                 animation_frame='Year',
                 animation_group='Country'
                 )

# Set bar chart layout
fig_bar.update_layout(title='Percentage road and rail passenger km in millions',
                      yaxis_title='Percentage',
                      legend_title_text='',
                      )

# Set hovertemplate
fig_bar.update_traces(hovertemplate='<b>Country:</b> %{x}<br>'
                                    '<b>Percentage:</b> %{y:,.1f}'
                                    '<extra></extra>')

# Show bar chart
fig_bar.show()

Overall, we notice that eastern European countries make more use of buses and coaches compared to western European countries. Of course, this does depend on multiple factors.
1. Environment:
    Is the country hilly/mountainous or is it flat? Is there space in cities for buses/coaches or is an underground rail network a better option?
2. Time:
    One can assume it is easier and cheaper on the short term to add buses and coaches in the travel infrastructure than it is to add a rail network. However, on the long term, a rail network could bring in more people due to its efficiency and potential to be controlled remotely.
3. GDP:
    The income of a country also affects what infrastructure can be build. Typically, a rail network is more expensive than creating a bus network. However, the data shows that Italy, which we defined as a low income country, uses pretty much only rail travel when compared to bus travel. And Finland, which we defined as a high income country, uses more road travel. This of course could be because of other factors. 

In [18]:
# Create line graph to disaplay the trend of rail passenger transport
country_name = input(f"Country name")

if country_name in df_rail_pg.index:
    # Obtain the selected country's data
    country_data = df_rail_pg.loc[country_name]
    # Obtain year as x-axis label
    years = country_data.index
    
    ## Obtain data as y-axis data
    data = country_data.str.replace(',', '').astype(int)
    
    #Remove commas and convert data to floating-point numbers while handling missing values
    filled_data = data.reindex(years).interpolate(method='linear')
    
    # Create line chart
    fig, ax = plt.subplots()
    ax.plot(years, data, marker='o', linestyle='-')
    ax.set_xlabel('Year')
    ax.set_ylabel('Millions of passenger-kilometres')
    ax.set_title(f'Rail transport of passengers in {country_name}')
    ax.grid(True)
    
    # Add data value labels
    for i, j in zip(years, data):
        ax.annotate(f"{j:,d}", xy=(i, j), xytext=(5,5), textcoords='offset points', fontsize=8, ha='center')
    
    ax.set_yticks(range(int(min(data)), int(max(data) + 1), 10000))
    plt.xticks(rotation=100)
    plt.show()
else:
    print(f"There's no data of {country_name}")

There's no data of France
